core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -174,23 +183,25 @@ class Replica(val brokerId: Int,
    * to the high watermark if there are no transactional messages in the log. Note also that the LSO cannot advance
    * beyond the high watermark.
    */
-  def lastStableOffset: LogOffsetMetadata = {
+  def lastStableOffsetMetadata: LogOffsetMetadata = {
     log.map { log =>
       log.firstUnstableOffset match {
-        case Some(offsetMetadata) if offsetMetadata.messageOffset < highWatermark.messageOffset => offsetMetadata
-        case _ => highWatermark
+        case Some(offsetMetadata) if offsetMetadata.messageOffset < highWatermark => offsetMetadata
+        case _ => highWatermarkMetadata
       }
     }.getOrElse(throw new KafkaException(s"Cannot fetch last stable offset on partition $topicPartition's " +
       s"non-local replica $brokerId"))
   }
 
+  def lastStableOffset: Long = lastStableOffsetMetadata.messageOffset
+
   /*
    * Convert hw to local offset metadata by reading the log at the hw offset.
    * If the hw offset is out of range, return the first offset of the first log segment as the offset metadata.
    */
   def convertHWToLocalOffsetMetadata() {
     if (isLocal) {
-      highWatermarkMetadata = log.get.convertToOffsetMetadata(highWatermarkMetadata.messageOffset).getOrElse {
+      _highWatermarkMetadata = log.get.convertToOffsetMetadata(_highWatermarkMetadata.messageOffset).getOrElse {
```

Comment:
I know this is not introduced in this PR, but the logic is a bit confusing to me: why read the metadata by getting the current metadata's offset, and then reading one byte from the log and reassign to the metadata again? It seems we only need to do anything if `_highWatermarkMetadata` is not defined, since otherwise we just unnecessarily read out the same _highWatermarkMetadata from the log.

Also nit: convertToOffsetMetadata(highWatermark).getOrElse

Reply:
Yep, that makes sense.